Compare metadata written by sim_telarray with array model parameters.

GernotMaier opened the pull request:
Add functionality to compare the metadata written by sim_telarray into its data file with values used to write the sim_telarray configuration files. This tests that the configuration has been correctly propagated to sim_telarray and used in the correct way.

In detail, this PR adds:

- a new module called `simtools.simtel.simtel_io_metadata` with two main functions:
  - read metadata from sim_telarray file
  - allow to get the sim_telarray telescope ID for a given CTA telescope naming (e.g., LSTN-02 to '2')
- a new module called `simtools.testing.sim_telarray_metadata` which compares for site and telescope parameter the values of a given model with those read from the metadata file

Closes #456

orelgueta commented on src/simtools/testing/sim_telarray_metadata.py:
```diff
@@ -0,0 +1,164 @@
+
+    telescope_parameter_mismatch = [
+        _assert_model_parameters(telescope_meta[i], model)
+        for i, (_, model) in enumerate(array_model.telescope_model.items(), start=1)
```
Probably worth replacing with `for i, model in enumerate(array_model.telescope_model.values(), start=1)`

GernotMaier replied:
OK, done!

orelgueta commented on src/simtools/testing/sim_telarray_metadata.py:
```diff
@@ -0,0 +1,164 @@
+        for i, (_, model) in enumerate(array_model.telescope_model.items(), start=1)
+    ]
+
+    # ensure printout of all mismatches, not only those found first
```
This comment is true for the second `if` statement, but if there's a mismatch in the site parameters, we won't reach the telescope parameters. Either we collect all errors and raise the ValueError afterwards, or perhaps join the mismatch lists so the `any` applies to the site one as well? Otherwise I don't mind if you simply want to move this comment to above the telescope mismatch check, at least there it will be correct.

GernotMaier replied:
I am aware of that. Your suggestion is good to merge the site / telescope mismatches and then check them. This allows to print all errors.

orelgueta commented on src/simtools/testing/sim_telarray_metadata.py:
```diff
@@ -0,0 +1,164 @@
+                )
+            else:
+                value = metadata[sim_telarray_name]
+                value = (int)(value) if value.isnumeric() else value
```
Why is it always cast to `int` here? Don't we have some floats in the configuration?

orelgueta replied:
OK, that is indeed a bit confusing (and made me think of negative integers as well), but this code works, so no complaints.

orelgueta commented on src/simtools/testing/sim_telarray_metadata.py:
```diff
@@ -0,0 +1,164 @@
+
+    Parameters
+    ----------
+    value1: any
+        First value to compare.
+    value2: any
+        Second value to compare.
+    value_type: str
+        Type of the values ('string', 'dict', etc.).
+
+    Returns
+    -------
+    bool
+        True if the values are equal, False otherwise.
+    """
+    value1 = value1[0] if isinstance(value1, tuple) else value1
+    value2 = value2[0] if isinstance(value2, tuple) else value2
+    if value1 is None or value2 is None:
+        if value1 in ("none", None) and value2 in ("none", None):
+            return True
+    if value_type == "string":
+        return str(value1).strip() == str(value2).strip()
+    if value_type == "dict":
+        return value1 == value2
+    if value_type == "boolean":
+        return bool(value1) == bool(value2)
+    return _is_equal_floats_or_ints(value1, value2)
+
+
+def _is_equal_floats_or_ints(value1, value2):
+    """Check if floats and ints are equal."""
+    if isinstance(value1, np.ndarray | list) and isinstance(value2, np.ndarray | list):
+        return bool(np.allclose(np.array(value1), np.array(value2), rtol=1e-10))
+    if isinstance(value1, list) and isinstance(value2, float | int | np.integer | np.floating):
+        if all(x == value1[0] for x in value1):
+            return bool(np.isclose(float(value1[0]), float(value2), rtol=1e-10))
+    if isinstance(value2, list) and isinstance(value1, float | int | np.integer | np.floating):
+        if all(x == value2[0] for x in value2):
+            return bool(np.isclose(float(value1), float(value2[0]), rtol=1e-10))
+    return bool(np.isclose(float(value1), float(value2), rtol=1e-10))
```
We could consider moving these two functions to general utilities (in particular if the rtol is made an argument).

GernotMaier replied:
Yes and no - this is a bit fine tuned to the model parameters (e.g., the conversion from json types to python types), the "none", "None" conversion, the possibility for tuble input) and the style we store the data (numpy arrays). 

Do you have another application in mind where this might be useful?

orelgueta replied:
No, no application, it just looked like a fairly general function (in particular the second one). I really don't mind it staying here though, so I am resolving this comment.

orelgueta commented on tests/unit_tests/simtel/test_simtel_io_metdata.py:
```diff
@@ -0,0 +1,37 @@
+    )
+
+
+def test_read_sim_telarray_metadata(test_sim_telarray_file):
```
Is it worth also testing that some random values in the file are as expected?

orelgueta commented on tests/unit_tests/testing/test_sim_telarray_metadata.py:
```diff
@@ -0,0 +1,138 @@
+    return read_sim_telarray_metadata(test_metadata_file)
+
+
+def test_assert_sim_telarray_metadata_success(test_metadata_file):
+    """Test assert_sim_telarray_metadata with successful comparison."""
```
This test is not for a success case, right?

GernotMaier replied:
Correct. Fixed this.

orelgueta reviewed (commented):
This is very nice and simple. A few minor comments.

GernotMaier commented on src/simtools/simtel/simtel_io_metadata.py:
```diff
@@ -0,0 +1,88 @@
+    }
+
+
+def get_sim_telarray_telescope_id(telescope_name, file):
```
I have asked around (AI) and the answer is consistent with what we discussed in person.

```
I would recommend removing the @cache decorator from this function since the expensive file I/O operation is already cached by the parent function.
````

On the bidirectional map: I would rather wait until we need it. Don't think we should implement code because it might be useful at some point. Will be quick to update it when needed.

orelgueta replied:
Yes, as we discussed in person, I agree with both points.

GernotMaier commented:
Thanks for the review @orelgueta . I've addressed most of your suggestions - there are two items which might need a bit more discussions. Let me know if this should be part of this pull request.

orelgueta commented on src/simtools/testing/sim_telarray_metadata.py:
```diff
@@ -40,20 +40,15 @@ def assert_sim_telarray_metadata(file, array_model):
-        )
+    telescope_parameter_mismatch.append(site_parameter_mismatch)
     if any(len(m) > 0 for m in telescope_parameter_mismatch):
+        mismatches = [m for m in telescope_parameter_mismatch if len(m) > 0]
```
Is this needed now? It wasn't needed before as far as I understood (i.e., it was printing the errors for all of them one by one before as well).

GernotMaier replied:
This line cleans up the output. Otherwise empty list brackets `[]` are printed for each telescope. Simply to make the error message easier to read.

orelgueta reviewed (approved):
All good now, thanks! I added one question but it's not blocking of course, so I am approving.

Not sure which two points require discussion, it seems we agree on everything already.
